### PR TITLE
Fix: controlled rounding for sub-sub-normals

### LINKED ISSUE #1997: Dashu FBig `to_f32` and `to_f64` don't always preserve rounding mode
Specifically when an arbitrary-precision float is positive, but smaller than the minimum subnormal in the target native type, the cast to the target native type with controlled rounding up still outputs zero. 

In an extreme case, in the upcoming refactored zCDP to approxDP code, this would result in an underflow of delta to zero when epsilon is ~9 and rho is 0.5. Therefore the underflow would cause a mis-representation of the privacy guarantee: if delta is zero, then epsilon should be infinity, not ~9.

### PR DESCRIPTION
**Pull request by Shoeboxam**
Fixes #1997

I'd consider this more of a patch than a fix. At some point when Dashu is updated, one of our tests will start to fail, and we can try to remove the patch.

https://github.com/cmpute/dashu/issues/53

### CODE REVIEW

**Comment by Shoeboxam** [comment]
* **#1768** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1768?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#1854](https://github.com/opendp/opendp/pull/1854) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1854?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#1999](https://github.com/opendp/opendp/pull/1999) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1999?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1998** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1998?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by mccalluc — COMMENTED** [review]
To confirm, the only test that _shouldn't_ pass is `test_subnormal_fail`? The others just improve the coverage, and would stay around, even if the other code is reverted.

Just glanced at the actual code changes: Let me know if there's any part I should invest more time in understanding?

A little concerned about making sure it's easy to make the necessary reversions when dashu is updated. Maybe it's enough to point to this URL from the test so future you can see the changes easily, or many we want a little comment on every block, so we don't end up inadvertantly making more changes there?

- Would you like to either keep the issue open or file a new issue to remind us to check if dashu has fixed this and to upgrade?
- Would a note in `Cargo.toml` on the dashu line be useful? I'm imagining the upstream bug gets fixed, but we don't immediately notice, and upgrade dashu for some other reason. But maybe the note in the test is sufficient.
- Will all the changes in `cast/mod.rs` be reverted at that point? Maybe a note on every block to remind us that this code is going away?

File: rust/src/traits/cast/mod.rs [hunk 1]
```diff
@@ -236,20 +237,177 @@ impl ExactIntBounds for f32 {
...
 }
 
-impl<R: Round> FromFBig<R> for f32 {
+impl<R: Round> FromFBig<R> for f32
+where
+    FBig<R>: ToFloatRounded,
+{
     fn from_fbig(value: FBig<R>) -> Self {
-        value.to_f32().value()
+        value.to_f32_rounded()
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/traits/cast/mod.rs:401]: I think the test is sufficient. Added another note in the test.

**Comment by Shoeboxam** [comment]
The test is meant to be a canary for when Dashu behavior changes, and I've updated it with your comments.

The code for taking the next float or previous float will also be used in the upcoming approximate zCDP key release.

**Review by mccalluc — APPROVED** [review]
Thanks!

**Pull request merged**